### ISSUE TITLE
Keyboard navigation on popup elements

### ISSUE THREAD

**Issue by Bloke** [textpattern/textpattern#903]
Keyboard navigation of some Ajax-driven popup elements is sub-optimal or broken, and success is browser dependent.

Write panel, STR:
* Navigate to Textfilter choice using tab key.
* Open it up (Enter key?).
* Try to select an option using the keyboard.

Articles panel, STR:
* Navigate to either list options or Search options dropdowns.
* Open up the list using Enter.
* Try to select an option using the keyboard.

Expected outcome: popup option lists should be keyboard navigable and selectable.

See also https://github.com/textpattern/textpattern/issues/887#issuecomment-295015554

**Comment by phiw13** [comment]
Follow up on my [comment](https://github.com/textpattern/textpattern/issues/887#issuecomment-295015554), it does work fine on touch devices (at least iOS powered devices). I don’t know about screen readers, I have a hard time testing that as I can barely hear what VoiceOver is telling me (the state of my hearing being rather poor…).

**Comment by philwareham** [comment]
@Bloke Hey Stef, regarding the write panel textfilter UI, I'm going to need the HTML structure changed to the following expectation:

    <div class="txp-textarea-options txp-textfilter-options no-ui-button">
        <label>Format:
        <select class="jquery-ui-selectmenu">
            <option>Plain text</option>
            <option>Plain text with linebreaks</option>
            <option selected>Textile</option>
            <option>Markdown (Parsedown)</option>
        </select></label>&#160;<a role="button" rel="help" target="_blank" class="pophelp" title="Help" aria-label="Help" href="#">i</a>
    </div>

That will improve that UI widget and make it keyboard accessible.

For the articles panel it's not so simple, I'll have to investigate that further (the checkboxes within being the tricky element).

**Comment by philwareham** [comment]
Bumping this, as I can't do it directly due to lack of PHP chops.

**Comment by Bloke** [comment]
I've got the code stashed. Dunno why I never finished it. I'll unstash it later and check.

**Comment by Bloke** [comment]
@phiw13 Please test d29bd0f0 and let me know if it's any better. It seems to be more keyboard navigable for me, and a lot less code! Note that your `gTxt()` textfilter string may still have a `{filter}` parameter in it. This isn't required any more so the string will need updating on the server at some point.

**Comment by phiw13** [comment]
@Bloke Works great with Safari and Firefox on my Mac. I’ll try later to test with Edge on Win 10 and iOS.

**Comment by philwareham** [comment]
Great, so we just need to investigate the search options popup now, I need to think about that as it's not an easy fix, due to the checkboxes inside a jQuery UI menu. Maybe a `tabindex: -1` is needed on the list items (but not the checkboxes) or something. Don't know yet - will let you know.

**Comment by Bloke** [comment]
Yes, not sure about those. We might have to enlist some cunning @bloatware JS, because making the search options and list-options keyboard navigable is probably above my pay grade.

**Comment by philwareham** [comment]
OK, as a follow up to this, we'll probably need to use something akin to this: https://github.com/ehynds/jquery-ui-multiselect-widget and change the menus with checkboxes (search options and list column options) to multi-select tags in HTML. Will investigate further.

**Comment by philwareham** [comment]
Or something using the existing checkboxradio UI we have available: http://design-patterns.textpattern.io/docs/#ui-doc-jquery-ui--checkboxradio. Hmmm.

**Comment by bloatware** [comment]
FWIW, I (Firefox W7) *can* navigate menus with tab key... do I miss something?

**Comment by Bloke** [comment]
STR:

* Go to Content->Articles panel.
* Tab to the Search box options (the down arrow on the for right).
* Hit space bar to display checkbox list.
* Hit 'Tab' to enter the checkbox list (though down-arrow should work, right?)
* Use arrow keys to go up and down the checkbox options.
* Try to select one - Enter? Spacebar? Left/right arrows? Nothing works.

**Comment by philwareham** [comment]
FYI The menus work a little better in Firefox, at least you can keyboard select checkboxes on/off. It's totally broken in Chrome and Safari though.

**Comment by phiw13** [comment]
And, as I noted in my [original comment](https://github.com/textpattern/textpattern/issues/887#issuecomment-295015554) it is totally broken in Edge (and IE 11 if memory serves). FWIW, it never worked really well in Firefox.

**Comment by bloatware** [comment]
Tab and Shift+Tab (plus Spacebar) work for me, the arrows don't. But there is no reason why they would - there is no standard behavior AFAIK.

**Comment by Bloke** [comment]
> Tab and Shift+Tab (plus Spacebar)

Weird, doesn't work for me in FF 54(!) on Win7. If I tab into the optionsl list, I can scroll up and down the checkbox values, but I can't find any keystroke that (de)activates a checkbox. Tried spacebar, Enter, 'x', left/right arrow keys, '+', '-', and so on. Nothing.

**Comment by bloatware** [comment]
Same for regular (ie non JQueryUI) checkboxes? Then we are not guilty :-)

**Comment by philwareham** [comment]
No, it's definitely broken. I am investigating solutions and will get back to you all.

**Comment by phiw13** [comment]
> Weird, doesn't work for me in FF 54(!) on Win7

Worked for me sort of on win 10 this afternoon (on the search pop-over, bu not on the option pop-over)

> Same for regular (ie non JQueryUI) checkboxes?

tab + spacebar works fine for no jq-ui checkboxes on every browser I ever tried.

**Comment by philwareham** [comment]
tab+space is a globally-accepted way of activating checkbox/radio buttons so it should be respected here too. 

Right, using a jQuery UI plugin as mentioned above I've put together a proof of concept - forget that it's not pretty right now but it allows what we need: http://design-patterns.textpattern.io/mockups/content-list-test.html

The checkboxes will have to be replaced in the system with a multi-select tag but that is still semantically OK. Let me know your thoughts?

**Comment by Bloke** [comment]
As long as the multi-select tool fires events that we can trap, it should be okay. We need those to be able to fire localStorage events and trigger on-page (and Ajax) behaviour. We also need shift-click behaviour across checkbox sets, because this impacts our multi-select feature. I'd rather not implement a new function suite just for these 'faux' checkbox types.

OT: is that the default multi-select styling in jQuery UI? Looks like a dog's breakfast! Mind you, I've always hated their default 'tabs' implementation too. Soooooo ugly.

**Comment by phiw13** [comment]
> Right, using a jQuery UI plugin as mentioned above I've put together a proof of concept (see search widget) - forget that it's not pretty right now but it allows what we need: http://design-patterns.textpattern.io/mockups/content-list-test.html

Hmm, not really intuitive that you have to arrow down to select the next checkbox instead of the common tab press. But it actually works, once you know it.

**Comment by philwareham** [comment]
@phiw13 As far as I'm aware, arrow key navigation is standard for `select` elements. I know this UI gives the impression of checkboxes instead, but it's better than it being totally inaccessible to keyboard. I'm OK with it.

@Bloke No it's totally unstyled by CSS at the moment (including the width), apart from what it's inherited from our form tags. I don't see how this `select` tag this affects the checkboxes used on lists themselves - they are not related at all to this widget and don't use the same code or HTML tags.

**Comment by philwareham** [comment]
Also, it you are using multi-`select` tag in other places but don't want to use jQuery UI for those, simply don't put the classname hook on them - they'll continue to function as vanilla multi-`select`.

**Comment by phiw13** [comment]
> As far as I'm aware, arrow key navigation is standard for select elements.

Correct, yes.

> I know this UI gives the impression of checkboxes instead, but it's better than it being totally inaccessible to keyboard. I'm OK with it.

As I said, it actually works yes, assuming to sends the events required. This calls for a `-moz-appearance: none; -webkit-appearance:none` on those checkboxes and use a custom marker. Hmmm.

**Comment by Bloke** [comment]
> I don't see how this select tag this affects the checkboxes used on lists themselves - they are not related at all to this widget and don't use the same code or HTML tags.

At the moment, we use our internal `txpMultiEditForm()` JS code to create the checkboxes in both the [search](https://github.com/textpattern/textpattern/blob/dev/textpattern/textpattern.js#L1533) and [columniser](https://github.com/textpattern/textpattern/blob/dev/textpattern/textpattern.js#L1623). The fact we can't shift-click items there is down to some random bug I never fathomed. In theory, it supports it.

Now, if we're changing from multi-edit checkboxes to a multi-select input widget that _looks like_ checkboxes I was just wondering if we could refactor `txpMultiEditForm()` to support both, instead of writing a new handler. But I guess if it's a multi-select list, then ctrl/shift click comes as standard so maybe a new handler is the right way to go.

**Comment by philwareham** [comment]
> ... and use a custom marker. Hmmm.

We already have a custom marker available (see http://design-patterns.textpattern.io/docs/#ui-doc-jquery-ui--checkboxradio) so that's certainly doable.

**Comment by Bloke** [comment]
So what needs doing here now? Which controls need revamping to use the new library that's been included? And what do they need to look like in terms of markup?

**Comment by philwareham** [comment]
It needs us to finish the integration of jquery-ui-multiselect-widget. I should be able to do my part of that (styling and demo mockup) soon. Been a busy summer but hopefully will get some time to work on Textpattern after next couple of weeks.

**Comment by Bloke** [comment]
Cool, no worries. Just wanted to check you weren't waiting on me to do something here!

**Comment by philwareham** [comment]
Please try the proposed fix for this in my flat mockup (specifically the search options dropdown) using the jquery.multiselect.js jQuery UI plugin...

http://design-patterns.textpattern.io/mockups/content-list.html

...if that is satisfactory I'll work up the list columns options dropdown into the same widget and start hooking it up for real in the Textpattern `issue-903` branch. Basically it uses a multiselect HTML tag to achieve this (which is semantically correct too).

**Comment by Bloke** [comment]
Looks great and works way better than other systems we've tried.

It is kinda weird that you need to tab twice to get to the checkboxes. At least in Firefox.

First tab takes you to the list items themselves. You can arrow up/down or use keyboard shortcuts of the first letter to jump to an item, but can't select anything until you tab again (which takes you to the top checkbox) then you can move up and down the checkboxes themselves using tab/shift-tab and select/deselect with the space bar.

In order to close the box you need to shift-tab all the way "out" to the List Options and hit enter. Is that normal/expected behaviour?

**Comment by phiw13** [comment]
Haven’t tried in Firefox yet, but with Safari: you press enter / spacebar on the arrow button (tab 2x from the search box to focus), the menu opens and the top checkbox is focussed. Press spacebar to toggle, or arrow down to move to another item (focus ring follows on the checkbox), rinse & repeat to toggle. The one issue with this is closing the box: you need to press tab to close it –the focus returns on the button– then shift tab to to the magnifier icon to search.

But it works much better than the previous popup, that is sure.

**Comment by philwareham** [comment]
You can press escape to exit the menu, same as for our navigation menus I believe.

**Comment by phiw13** [comment]
yes `esc` works to move out, but conceptually it is associated with “cancel”…

Also, wonder how this comes over: https://dev.l-c-n.com/txp/txp-content-list-articles-multiselect-dropdown.html

as is it works fine in most browsers, IE 11 has an issue (easily fixed). Those are fully styled checkboxes, hidden with `opacity`, visible when in `:checked` state.

(server OS is being upgraded, may be not accessible over the next couple of hours)

![multi-select](https://user-images.githubusercontent.com/8944809/32865924-9f7becfa-caa8-11e7-894e-64691ba80437.png)

**Comment by phiw13** [comment]
Hmmm. Firefox 57 (both macOS and Win10-build1709): from the search text field tab 2x, press spacebar the popup opens (or click on the button to open). Arrow key down, press spacebar to toggle the checkbox. It toggles briefly between `checked` and off, but it doesn’t keep that state. Using the mouse works great. All other browsers work as expected.

**Comment by phiw13** [comment]
Ignore the previous comment. Firefox **does** work but one has to press the `enter` or `return` key to toggle the checkboxes (something that works with Safari as well, prolly others). In that case I’ll have review  my design idea.

**Comment by Bloke** [comment]
I couldn't get `esc` to work on Firefox 57 (Mac). Nor did Enter on the checkboxes. Space works though.

But if there's a way to navigate it via keyboard on all browsers - even if the mechanisms to do so are inconsistent - then I say we print this and move on.

**Comment by philwareham** [comment]
OK I’ll start putting this into the branch. Might need done help with the PHP changes but I’ll let you know.

**Comment by phiw13** [comment]
@Bloke - I have no idea why you see a different behaviour than I do with Firefox. The only difference that I can think about, and this is a huge, massive stretch, is that you have a different keyboard than I have (I have a JIS keyboard, while yours is probably the British one? my Win10 laptop also has a Jpn keyboard). BTW, I noticed that while the popup is open, the script blocks many more standard keyboard shortcuts than Safari, Cmd-H for example.

Anyway, accessibility wise, this is a huge improvement over the current situation. Ship it!

(if I tested correctly it will solve this problem: https://github.com/philwareham/textpattern-hive-admin-theme/issues/103)

——

@philwareham  - looking at the latest version of the mockup, the list-options button is kinda odd - it looks a bit like a zombie button with no clear explanation. I think it would be clearer with the label fully visible, see my mockup: https://dev.l-c-n.com/txp/txp-content-list-articles.html

**Comment by philwareham** [comment]
@phiw13 I think you may be caching an older JS / the list options label should already be visible. Let me know?

**Comment by phiw13** [comment]
On http://design-patterns.textpattern.io/mockups/content-list.html, I see this:

![hive2017-11-17](https://user-images.githubusercontent.com/8944809/32935464-a845b9c2-cbb3-11e7-8090-7cc2a62226eb.png)

Should I see more? I specifically refer to show the text next to the icon, to make the widget more understandable.

**Comment by philwareham** [comment]
Oh, ok, maybe I forgot to upload my latest code to live. Yes the label should be there.

**Comment by philwareham** [comment]
Phew! We got there in the end - closing this bloody issue!! Thanks everyone.
